client/rest-high-level/src/main/java/org/elasticsearch/client/security/user/privileges/ApplicationResourcePrivileges.java

```diff
@@ -0,0 +1,182 @@
+                final Collection<String> privileges = (Collection<String>) constructorObjects[i++];
+                final Collection<String> resources = (Collection<String>) constructorObjects[i];
+                return new ApplicationResourcePrivileges(application, privileges, resources);
+            });
+
+    static {
+        PARSER.declareString(constructorArg(), APPLICATION);
+        PARSER.declareStringArray(constructorArg(), PRIVILEGES);
+        PARSER.declareStringArray(constructorArg(), RESOURCES);
+    }
+
+    private final String application;
+    private final Set<String> privileges;
+    private final Set<String> resources;
+
+    private ApplicationResourcePrivileges(String application, Collection<String> privileges, Collection<String> resources) {
+        // we do all null checks inside the constructor
+        if (Strings.isNullOrEmpty(application)) {
+            throw new IllegalArgumentException("application privileges must have an application name");
+        }
+        if (null == privileges || privileges.isEmpty()) {
+            throw new IllegalArgumentException("application privileges must define at least one privilege");
+        }
+        if (null == resources || resources.isEmpty()) {
+            throw new IllegalArgumentException("application privileges must refer to at least one resource");
+        }
+        this.application = application;
+        this.privileges = Collections.unmodifiableSet(new HashSet<>(privileges));
+        this.resources = Collections.unmodifiableSet(new HashSet<>(resources));
+    }
+
+    public String getApplication() {
+        return application;
+    }
+
+    public Set<String> getResources() {
+        return this.resources;
+    }
+
+    public Set<String> getPrivileges() {
+        return this.privileges;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || this.getClass() != o.getClass()) {
+            return false;
+        }
+        ApplicationResourcePrivileges that = (ApplicationResourcePrivileges) o;
+        return application.equals(that.application) && privileges.equals(that.privileges) && resources.equals(that.resources);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(application, privileges, resources);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder(getClass().getSimpleName()).append("[");
+                sb.append(APPLICATION.getPreferredName()).append("=[").append(application).append("], ");
+                sb.append(PRIVILEGES.getPreferredName()).append("=[").append(Strings.collectionToCommaDelimitedString(privileges)).append("], ");
+                sb.append(RESOURCES.getPreferredName()).append("=[").append(Strings.collectionToCommaDelimitedString(resources)).append("]]");
+        return sb.toString();
```

Comment:
There are objects that don't implement `toXContent` (because they are not part of a request) but implement `toString`. Would they keep their present `toString` format or use the JSON format. If they keep the present format, then the `toString` will not be consistent across the project (some return JSON others don't). If every `toString` returns JSON then everyone implements `ToXContentObject` and we need to test for that.

I would keep it as is, it's not that we add entities everyday, just this specific API is rich in this sort of objects.

Reply:
I have replaced the `toString` with the JSON format. I still believe having objects with mixed formats for `toString` is messy (objects with non-JSON format can contain objects with JSON format) and imposing JSON format for everyone is impractical. But I might be weighing in trade-offs wrong.

Reply:
I certainly didn't mean to imply that everything needed a JSON format. Just when a class already has a toXContent method, maintaining a separate toString (and making sure it gets updated with each new field) seems like wasted effort.
Sometimes there's an argument that the JSON is too verbose, but I'd prefer we make the argument when it needs to be different rather than just keep writing & mtaining both as a default.